RecoVertex/PrimaryVertexProducer/src/DAClusterizerInZT.cc

```diff
@@ -0,0 +1,778 @@
+      + (std::pow(beamspot.BeamWidthX()*t_mom.x(),2)+std::pow(beamspot.BeamWidthY()*t_mom.y(),2))*std::pow(t_mom.z(),2)/std::pow(t_mom.perp2(),2) // beam spot width
+      + std::pow(vertexSize_, 2); // intrinsic vertex size, safer for outliers and short lived decays
+    t.dz2 = 1./ t_dz2;
+    
+    t.dtz = 0.;
+    t.dt2 = std::pow((*it).dtErrorExt(),2.0) + std::pow(0.010,2.0);   // the ~injected~ timing error, need to add a small minimum vertex size in time
+    
+    if (d0CutOff_>0){
+      Measurement1D IP=(*it).stateAtBeamLine().transverseImpactParameter();// error constains beamspot
+      t.pi=1./(1.+std::exp(std::pow(IP.value()/IP.error(),2.0) - std::pow(d0CutOff_ ,2)));  // reduce weight for high ip tracks  
+    }else{
+      t.pi=1.;
+    }    
+    t.tt=&(*it);
+    t.Z=1.;
+    if( edm::isFinite(t.pi) &&  t.pi >= std::numeric_limits<double>::epsilon() ) {
+      tks.push_back(t);
+    }
+  }
+
+  return tks;
+}
+
+
+
+
+
+double DAClusterizerInZT::Eik(const track_t & t, const vertex_t &k ) const {
+  return std::pow(t.z-k.z,2.)/t.dz2 + std::pow(t.t - k.t,2.)/t.dt2;
+}
+  
+
+
+
+double DAClusterizerInZT::update( double beta,
+                                  vector<track_t> & tks,
+                                  vector<vertex_t> & y ) const {
+  //update weights and vertex positions
+  // mass constrained annealing without noise
+  // returns the squared sum of changes of vertex positions
+
+  unsigned int nt=tks.size();
+
+  //initialize sums
+  double sumpi=0;
+  for(vector<vertex_t>::iterator k=y.begin(); k!=y.end(); ++k){
+    k->sw=0.; k->swz=0.; k->swt = 0.; k->se=0.;
+    k->swE=0.;  k->Tc=0.;
+  }
+
+
+  // loop over tracks
+  for(unsigned int i=0; i<nt; i++){
+
+    // update pik and Zi
+    double Zi = 0.;
+    for(vector<vertex_t>::iterator k=y.begin(); k!=y.end(); ++k){
+      k->ei = std::exp(-beta*Eik(tks[i],*k));// cache exponential for one track at a time
+      Zi   += k->pk * k->ei;      
+    }
+    tks[i].Z=Zi;
+
+    // normalization for pk
+    if (tks[i].Z>0){
+      sumpi += tks[i].pi;
+      // accumulate weighted z and weights for vertex update
+      for(vector<vertex_t>::iterator k=y.begin(); k!=y.end(); ++k){
+	k->se  += tks[i].pi* k->ei / Zi;
+	const double w = k->pk * tks[i].pi* k->ei / ( Zi * ( tks[i].dz2 * tks[i].dt2 ) );
+	k->sw  += w;
+	k->swz += w * tks[i].z;
+        k->swt += w * tks[i].t;
+	k->swE += w * Eik(tks[i],*k);
+      }
+    }else{
+      sumpi += tks[i].pi;
+    }
+    
+
+  } // end of track loop
+
+
+  // now update z and pk
+  double delta=0;
+  for(vector<vertex_t>::iterator k=y.begin(); k!=y.end(); k++){
+    if ( k->sw > 0){
+      const double znew = k->swz/k->sw; 
+      const double tnew = k->swt/k->sw;
+      delta += std::pow(k->z-znew,2.) + std::pow(k->t-tnew,2.);
+      k->z  = znew;
+      k->t  = tnew;
+      k->Tc = 2.*k->swE/k->sw;
+    }else{
+      edm::LogInfo("sumw") <<  "invalid sum of weights in fit: " << k->sw << endl;
```

Comment:
how frequently is this this scope  reached?
If more than a few times per event, LogInfo should be turned to a LogDebug or suppressed by other means